### ISSUE TITLE
Run benchmark models get hanged when use grpc+gdr communication.

### ISSUE THREAD

**Issue by lalalapotter** [tensorflow/tensorflow#41764]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: 
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): r1.15
- Python version: python3.6
- Bazel version (if compiling from source): 0.26.0
- GCC/Compiler version (if compiling from source): 7.5.0
- CUDA/cuDNN version: CUDA10.1
- GPU model and memory: V100 32GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When I compile tensorflow r1.15 version from source code with the gdr support. The compile command line is as following.
`bazel build --config=opt --config=cuda --config=gdr //tensorflow/tools/pip_package:build_pip_package`
After I run the benchmark models, I get the process hang with the following output.

> WARNING:tensorflow:From /home/cengguang/zcg/lib/python3.6/site-packages/tensorflow_core/python/compat/v2_compat.py:68: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
Instructions for updating:
non-resource variables are not supported in the long term
2020-07-27 15:27:24.230627: I tensorflow/contrib/gdr/gdr_memory_manager.cc:220] RDMA server is listening on 10.0.9.2:10001
2020-07-27 15:27:24.230719: I tensorflow/contrib/gdr/gdr_memory_manager.cc:86] NUMA node for device: mlx5_0 is 0
2020-07-27 15:27:24.230763: I tensorflow/contrib/gdr/gdr_memory_manager.cc:250] Instrumenting CPU allocator(s)
2020-07-27 15:27:24.245919: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2400000000 Hz
2020-07-27 15:27:24.246089: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5f6a630 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-07-27 15:27:24.246112: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-07-27 15:27:24.248568: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-07-27 15:27:26.772045: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5f72cc0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2020-07-27 15:27:26.772110: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Tesla V100-SXM2-32GB, Compute Capability 7.0
2020-07-27 15:27:26.776200: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1639] Found device 0 with properties:
name: Tesla V100-SXM2-32GB major: 7 minor: 0 memoryClockRate(GHz): 1.53
pciBusID: 0000:89:00.0
2020-07-27 15:27:26.776621: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2020-07-27 15:27:26.780467: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2020-07-27 15:27:26.784188: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2020-07-27 15:27:26.784806: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2020-07-27 15:27:26.787654: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2020-07-27 15:27:26.788934: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2020-07-27 15:27:26.793107: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2020-07-27 15:27:26.795799: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1767] Adding visible gpu devices: 0
2020-07-27 15:27:26.795862: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2020-07-27 15:27:26.798030: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1180] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-07-27 15:27:26.798050: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1186]      0
2020-07-27 15:27:26.798057: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1199] 0:   N
2020-07-27 15:27:26.800718: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1325] Created TensorFlow device (/job:worker/replica:0/task:1/device:GPU:0 with 30584 MB memory) -> physical GPU (device: 0, name: Tesla V100-SXM2-32GB, pci bus id: 0000:89:00.0, compute capability: 7.0)
2020-07-27 15:27:26.806113: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:258] Initialize GrpcChannelCache for job worker -> {0 -> 10.0.9.2:10009, 1 -> localhost:10001}
begin a gdr server
2020-07-27 15:27:26.810425: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:365] Started server with target: grpc://localhost:10001
WARNING:tensorflow:From /home/cengguang/benchmarks/scripts/tf_cnn_benchmarks/convnet_builder.py:134: conv2d (from tensorflow.python.layers.convolutional) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.keras.layers.Conv2D` instead.
W0727 15:27:26.848110 139800816277312 deprecation.py:323] From /home/cengguang/benchmarks/scripts/tf_cnn_benchmarks/convnet_builder.py:134: conv2d (from tensorflow.python.layers.convolutional) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.keras.layers.Conv2D` instead.
WARNING:tensorflow:From /home/cengguang/zcg/lib/python3.6/site-packages/tensorflow_core/python/layers/convolutional.py:424: Layer.apply (from tensorflow.python.keras.engine.base_layer) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `layer.__call__` method instead.
W0727 15:27:26.850684 139800816277312 deprecation.py:323] From /home/cengguang/zcg/lib/python3.6/site-packages/tensorflow_core/python/layers/convolutional.py:424: Layer.apply (from tensorflow.python.keras.engine.base_layer) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `layer.__call__` method instead.
WARNING:tensorflow:From /home/cengguang/benchmarks/scripts/tf_cnn_benchmarks/convnet_builder.py:266: max_pooling2d (from tensorflow.python.layers.pooling) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.MaxPooling2D instead.
W0727 15:27:26.874744 139800816277312 deprecation.py:323] From /home/cengguang/benchmarks/scripts/tf_cnn_benchmarks/convnet_builder.py:266: max_pooling2d (from tensorflow.python.layers.pooling) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.MaxPooling2D instead.
WARNING:tensorflow:From /home/cengguang/benchmarks/scripts/tf_cnn_benchmarks/convnet_builder.py:408: dropout (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dropout instead.
W0727 15:27:26.954734 139800816277312 deprecation.py:323] From /home/cengguang/benchmarks/scripts/tf_cnn_benchmarks/convnet_builder.py:408: dropout (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dropout instead.
WARNING:tensorflow:From /home/cengguang/zcg/lib/python3.6/site-packages/tensorflow_core/python/ops/losses/losses_impl.py:121: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
W0727 15:27:27.005284 139800816277312 deprecation.py:323] From /home/cengguang/zcg/lib/python3.6/site-packages/tensorflow_core/python/ops/losses/losses_impl.py:121: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
WARNING:tensorflow:From /home/cengguang/benchmarks/scripts/tf_cnn_benchmarks/benchmark_cnn.py:2268: Supervisor.__init__ (from tensorflow.python.training.supervisor) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.MonitoredTrainingSession
W0727 15:27:27.223294 139800816277312 deprecation.py:323] From /home/cengguang/benchmarks/scripts/tf_cnn_benchmarks/benchmark_cnn.py:2268: Supervisor.__init__ (from tensorflow.python.training.supervisor) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.MonitoredTrainingSession
INFO:tensorflow:Running local_init_op.
I0727 15:27:28.615225 139800816277312 session_manager.py:500] Running local_init_op.
2020-07-27 15:27:28.770876: I tensorflow/contrib/gdr/gdr_memory_manager.cc:572] RDMA endpoint connected to rdma://10.0.9.2:10009
INFO:tensorflow:Done running local_init_op.
I0727 15:27:29.256539 139800816277312 session_manager.py:502] Done running local_init_op.
2020-07-27 15:27:29.479763: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2020-07-27 15:27:29.801486: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2020-07-27 15:27:31.455157: I tensorflow/contrib/gdr/gdr_memory_manager.cc:279] Accepted new RDMA connection

**Describe the expected behavior**
Expected behavior should run the code smoothly and exit, BUT get stucked.
**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
After compile the source code with gdr support, you can use the following launch scripts:
`export CUDA_VISIBLE_DEVICES='0'
export LD_LIBRARY_PATH=/usr/local/cuda/lib64/
python tf_cnn_benchmarks.py \
        --controller_host=controller_ip \
        --worker_hosts=worker1_ip,worker2_ip\
        --variable_update=collective_all_reduce \
        --all_reduce_spec=collective \
        --job_name=worker \
        --use_fp16 \
        --batch_size=128 \
        --num_gpus=1 \
        --model=alexnet \
        --task_index=0 \
        --force_gpu_compatible \
        --server_protocol=grpc+gdr 2>&1| tee output1.txt`
**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by lalalapotter** [comment]
@junshi15 , @shamoya , @byronyi  can you give me some suggestions?

**Comment by lalalapotter** [comment]
When I check the Mellanox mlx5 Linux Counters and Status Parameters, I find that the following counters increase when the training process is hanged.
Content of rnr_nak_retry_err is: 220 -> 282 still increasing
Content of out_of_buffer is: 220 -> 282 still increasing
You can check the counters using following scripts:
`COUNTERS=`ls /sys/class/infiniband/mlx5_0/ports/1/hw_counters/`
ARRAY=(${COUNTERS///})
for var in ${ARRAY[@]}
do
  echo "Content of $var is:"
  cat /sys/class/infiniband/mlx5_0/ports/1/hw_counters/$var
  echo ""
done`

**Comment by junshi15** [comment]
I have not tracked this piece of work for a while, maybe @shamoya , @byronyi are more up-to-date on this.

**Comment by shamoya** [comment]
Same here. also not a Mellanox employee anymore.
From a short glimpse of the description, looks like you're running on a lossy Ethernet fabric. 
Try moving to lossless ethernet configuration (global pause / PFC) and see it reproduces.

Cheers

**Comment by gowthamkpr** [comment]
@lalalapotter Did you try moving to lossless ethernet configuration (global pause / PFC) and were you still able to reprodce this issue? Thanks!

**Comment by lalalapotter** [comment]
@gowthamkpr Hi, I can still reproduce the issue, it seems the problem is not related to the ethernet configuration. Do you have any other suggestions? Or can you try the grpc+gdr protocol?

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
